cranelift/codegen/src/isa/x64/inst/args.rs

```diff
@@ -0,0 +1,434 @@
+            RM::M { addr } => addr.show_rru(mb_rru),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+// Some basic ALU operations.  TODO: maybe add Adc, Sbb.
+pub enum RMI_R_Op {
+    Add,
+    Sub,
+    And,
+    Or,
+    Xor,
+    Mul, // the signless, non-extending (N x N -> N, for N in {32,64}) variant
+}
+
+impl RMI_R_Op {
+    pub(crate) fn to_string(&self) -> String {
+        match self {
+            RMI_R_Op::Add => "add".to_string(),
+            RMI_R_Op::Sub => "sub".to_string(),
+            RMI_R_Op::And => "and".to_string(),
+            RMI_R_Op::Or => "or".to_string(),
+            RMI_R_Op::Xor => "xor".to_string(),
+            RMI_R_Op::Mul => "imul".to_string(),
+        }
+    }
+}
+
+impl fmt::Debug for RMI_R_Op {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.to_string())
+    }
+}
+
+#[derive(Clone, PartialEq)]
+// These indicate ways of extending (widening) a value, using the Intel
+// naming: B(yte) = u8, W(ord) = u16, L(ong)word = u32, Q(uad)word = u64
+pub enum ExtMode {
+    BL, // B -> L
+    BQ, // B -> Q
+    WL, // W -> L
+    WQ, // W -> Q
+    LQ, // L -> Q
+}
+
+impl ExtMode {
+    pub(crate) fn to_string(&self) -> String {
+        match self {
+            ExtMode::BL => "bl".to_string(),
+            ExtMode::BQ => "bq".to_string(),
+            ExtMode::WL => "wl".to_string(),
+            ExtMode::WQ => "wq".to_string(),
+            ExtMode::LQ => "lq".to_string(),
+        }
+    }
+
+    pub(crate) fn dst_size(&self) -> u8 {
+        match self {
+            ExtMode::BL => 4,
+            ExtMode::BQ => 8,
+            ExtMode::WL => 4,
+            ExtMode::WQ => 8,
+            ExtMode::LQ => 8,
+        }
+    }
+}
+
+impl fmt::Debug for ExtMode {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.to_string())
+    }
+}
+
+// These indicate the form of a scalar shift: left, signed right, unsigned
+// right.
+#[derive(Clone)]
+pub enum ShiftKind {
+    Left,
+    RightZ,
+    RightS,
+}
+
+impl ShiftKind {
+    pub(crate) fn to_string(&self) -> String {
+        match self {
+            ShiftKind::Left => "shl".to_string(),
+            ShiftKind::RightZ => "shr".to_string(),
+            ShiftKind::RightS => "sar".to_string(),
+        }
+    }
+}
+
+impl fmt::Debug for ShiftKind {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.to_string())
+    }
+}
+
+// These indicate condition code tests.  Not all are represented since not all
+// are useful in compiler-generated code.
+#[derive(Copy, Clone)]
+pub enum CC {
+    O,   //  0   overflow
+    NO,  //  1   no overflow
+    B,   //  2   <u
+    NB,  //  3   >=u
+    Z,   //  4   zero
+    NZ,  //  5   not zero
+    BE,  //  6   <=u
+    NBE, //  7   >u
+    S,   //  8   negative
+    NS,  //  9   not negative
+    L,   //  12  <s
+    NL,  //  13  >=s
+    LE,  //  14  <=s
+    NLE, //  15  >s
```

Comment:
```suggestion
    /// overflow
    O = 0,
    /// no overflow
    NO = 1,
    /// <u
    B = 2,
    /// >= u
    NB = 3,
    /// zero
    Z = 4,
    /// not zero
    NZ = 5,
    /// <=u
    BE = 6,
    /// >u
    NBE = 7,
    /// negative
    S = 8,
    /// not negative
    NS = 9,
    /// <s
    L = 12,
    /// >=s
    NL= 13, 
    /// <=s
    LE = 14,
    /// >s
    NLE = 15,
```

Reply:
Is this really a win?  IMO it wastes a lot of vertical screen space for minimal gain.  Sure, we can rustdoc `CC` itself, but I think the benefit of documenting the meaning of each individual item is not worth the loss of screen space.

Reply:
I heard screen space is free, so I'll take it!